### ISSUE TITLE
Override variables from extended class

### ISSUE THREAD

**Issue by GammaGames** [godotengine/godot#19113]
Say I have a script, `Entity.gd`. It looks like:

```python3
extends KinematicBody2D

var speed = 0
var health = 3

func _ready():
    connect("",  self, "signal")
```

Now I want to extend that script into `Enemy.gd`. I want to change its health and speed, in order to do that I have to set them in the `_ready` function like so:

```python3
extends 'res://Entity.gd'

func _ready():
    speed = 5
    health = 5
    ._ready()
```

This seems like a somewhat hacky way to set variables in your base class, having to set your variables in `_ready` and then calling the parent's function if necessary.

I am proposing a small change to allow setting variables in a similar way to the base class, just using the `onready` keyword. The documentation says (referring to storing references to nodes):  
>For this, GDScript has the onready keyword, that defers initialization of a member variable until _ready is called

So `onready` is already used for a similar purpose in its current state. This suggestion would allow `Enemy.gd` to be written like:

```python3
extends 'res://Entity.gd'

onready speed = 5
onready health = 5
```

It would let scripts look more consistent when extending base classes, if you wanted to change one of the base class's variables you could do it in the same area that it was done in both scripts. `_ready` will be called just like in the base script. 

Thoughts?

**Comment by GammaGames** [comment]
I found out you could use `_init`, so now both scripts look like:

`Entity.gd`:
```python3
extends KinematicBody2D

var speed = 0
var health = 3    

_init(_speed = 0, _health = 3).():
    speed =_speed
    health = _health
```

`Enemy.gd`:
```python3
extends 'res://Entity.gd'
_init().(5, 5):
    pass
```

Which seems better, but I still think setting variables above would look nicer. Maybe with an `oninit` keyword?

**Comment by vnen** [comment]
I thought I commented here already, but it seems it did not register... Let me try again.

> This seems like a somewhat hacky way to set variables in your base class, having to set your variables in _ready and then calling the parent's function if necessary.

This is not hacky, it's the way OOP works in pretty much every language. However, I believe `_ready()` is in a multilevel call, so the parent will be called automatically.

In my lost comment I said that you can use `_init()` pretty much in the same way. You don't need `_ready()` unless you need something from the tree.

The thing is, you shouldn't expect this to work:

```gdscript
extends 'res://Entity.gd'

speed = 5
health = 5
```

So adding an `onready` keyword there should not magically make it work. This is not used in any language I know of, and it's not GDScript business to invent new constructs IMO. `onready` is a simple convenience to avoid creating a function when you only need to set up a few variables (so you don't need to declare in one place and initialize in another).

I think such shorthand can be quite cryptic, because the difference between declaration and parent override is just a `var` keyword that can be easily missed. I understand the appeal but IMO it's more confusing than useful.

**Comment by GammaGames** [comment]
>This is not hacky, it's the way OOP works in pretty much every language.

It more felt hacky because I thought there should be a better way to change inherited variables, and using `_init` is a better way IMO.

>I think such shorthand can be quite cryptic, because the difference between declaration and parent override is just a var keyword that can be easily missed. I understand the appeal but IMO it's more confusing than useful.

Those were my concerns as well, and there aren't really any arguments against them. Setting the values in `_init` sounds like it would work well, I'll use that. Thank you!
